File: Java/ToyLanguageGUI_exam_final/src/domain/statements/CloseRFile.java
package domain.statements;

import domain.*;
import domain.expressions.Expression;
import exceptions.InterpretorException;
import utils.IHeap;

import java.io.BufferedReader;
import java.io.IOException;

public class CloseRFile implements Statement {
    private Expression expFile;
    public CloseRFile(Expression ef){
        expFile = ef;
    }


    public boolean exists(PrgState p, int i)
    {
        IFileTable<Integer,FileData> ft = p.getFileTable();
        for(FileData key : ft.getValues())
        {
            if(key.getFileDescriptor() != null)
                return true;
        }
        return false;
    }

    @Override
    public PrgState execute(PrgState p) {
        try {
            IHeap<Integer,Integer> h = p.getHeap();

            int value = expFile.eval(p.getSymbols(),h);
            if(!exists(p,value))
                throw new InterpretorException("File not found");
            BufferedReader buff = p.getFileTable().get(value).getFileDescriptor();

            buff.close();

            p.getFileTable().remove(value);
        }catch(IOException ioe){
            throw new InterpretorException(ioe.getMessage());
        }catch(RuntimeException e){
            System.out.println(e.getMessage());
        }


        return null;
    }

    @Override
    public String toString(){
        return "closeFile("+expFile+")";
    }
}

File: ToyInterpretor/ToyInterpretor/Domain/Statements/CloseRFile.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ToyInterpretor.Domain.Expressions;
using System.IO;

namespace ToyInterpretor.Domain.Statements
{
    class CloseRFile : Statement
    {
        private Expression expFile;
        public CloseRFile(Expression ef)
        {
            expFile = ef;
        }


        public Boolean exists(PrgState p, int i)
        {
            IFileTable<int, FileData> ft = p.getFileTable();
            foreach (FileData key in ft.getValues())
            {
                if (key.getFileDescriptor() != null)
                    return true;
            }
            return false;
        }

        
        public PrgState execute(PrgState p)
        {
            try
            {
                int value = expFile.eval(p.getSymbols());
                if (!exists(p, value))
                    throw new Exception("File not found");
                BufferedStream buff = p.getFileTable().get(value).getFileDescriptor();

                buff.Close();

                p.getFileTable().remove(value);
            }
            catch (IOException ioe)
            {
                throw new IOException(ioe.Message);
            }
            catch (Exception e)
            {
                Console.WriteLine(e.Message);
            }


            return p;
        }

        
        public override String ToString()
        {
            return "closeFile(" + expFile + ")";
        }
    }
}
